bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/link/AbstractLinkRegistry.java

```diff
@@ -44,45 +123,99 @@ protected AbstractLinkRegistry(final Class<P> providerClazz) {
+            final Set<L> forLinkedUID = linkedUidToLink.get(uid);
+            if (forItemName == null || forLinkedUID == null) {
+                return false;
+            } else {
+                return forItemName.parallelStream().anyMatch(forLinkedUID::contains);
             }
+        } finally {
+            toLinkLock.readLock().unlock();
         }
+    }
 
-        return false;
+    /**
+     * Returns if a link for the given item name exists.
+     *
+     * @param itemName item name
+     * @return true if a link exists, otherwise false
+     */
+    public boolean isLinked(final String itemName) {
+        toLinkLock.readLock().lock();
+        try {
+            return itemNameToLink.get(itemName) != null; // if present the set is not empty by definition
+        } finally {
+            toLinkLock.readLock().unlock();
+        }
+    }
+
+    /**
+     * Returns if a link for the given UID exists.
+     *
+     * @param uid UID
+     * @return true if a link exists, otherwise false
+     */
+    public boolean isLinked(final UID uid) {
+        toLinkLock.readLock().lock();
+        try {
+            return linkedUidToLink.get(uid) != null; // if present the set is not empty by definition
+        } finally {
+            toLinkLock.readLock().unlock();
+        }
     }
 
     /**
      * Returns the item names, which are bound to the given UID.
      *
      * @param uid UID
-     * @return a non-null collection of item names that are linked to the given UID.
+     * @return a non-null unmodifiable collection of item names that are linked to the given UID.
      */
     public Set<String> getLinkedItemNames(UID uid) {
-        Set<String> linkedItems = new LinkedHashSet<>();
-        for (AbstractLink link : getAll()) {
-            if (link.getLinkedUID().equals(uid)) {
-                linkedItems.add(link.getItemName());
+        toLinkLock.readLock().lock();
+        try {
+            final Set<L> forLinkedUID = linkedUidToLink.get(uid);
+            if (forLinkedUID == null) {
+                return Collections.emptySet();
             }
+            return forLinkedUID.parallelStream().map(link -> link.getItemName()).collect(Collectors.toSet());
+        } finally {
+            toLinkLock.readLock().unlock();
         }
-        return linkedItems;
     }
 
     /**
      * Returns all links for a given UID.
      *
      * @param uid a channel UID
-     * @return a set of links for the given UID
+     * @return an unmodifiable set of links for the given UID
      */
     public Set<L> getLinks(UID uid) {
-        Set<L> links = new LinkedHashSet<>();
-        for (L link : getAll()) {
-            if (link.getLinkedUID().equals(uid)) {
-                links.add(link);
-            }
+        toLinkLock.readLock().lock();
+        try {
+            final Set<L> forLinkedUID = linkedUidToLink.get(uid);
+            return forLinkedUID != null ? new LinkedHashSet<>(forLinkedUID) : Collections.emptySet();
```

Comment:
I know that `new LinkedHashSet<>(forLinkedUID)` is not really unmodifiable and we could create an unmovidiable view, but I assume we don't need to care. The caller should act as the collection is not modifiable and he will run into trouble if he modifies it (e.g. the empty collection). But if we return a modifiable (don't create a proxy instance) it should not matter.

Reply:
Any reason not to return a `HashSet` instead of `LinkedHashSet` or use `Collections.unmodifiableSet(forLinkedUID)`?

Reply:
> Any reason not to ...use Collections.unmodifiableSet(forLinkedUID)?

Yes, sure!
The `Collections.unmodifiableSet` creates an unmodifiable **view** of the collection.
So any change on the collection will affect the consumer of the view.
The consumer could run into `ConcurrentModificationException` if we change the collection while he uses the view.

Reply:
> Any reason not to return a `HashSet` instead of `LinkedHashSet`

Not really. `LinkedHashSet` has been used before and I did not change it. I haven't checked if read only access of a `HashSet` is more performant then of a `LinkedHashSet`.
If you think it is faster, then I will change.

Reply:
Ok make sense regarding not using `Collections.unmodifiableSet`. I would think `HashSet` has a slightly lower memory footprint and possible faster because no need to keep set linked. Since a generic Set is returned I would say there is no advantage to make it a `LinkedHashSet`.

Reply:
https://dzone.com/articles/java-collection-performance

![perfomance of set](http://4.bp.blogspot.com/-cNSGutQiqVw/UEFtS7SlHqI/AAAAAAAAAGI/0fSiC6Ok718/s1600/SetPerf.png)